esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+	w.buf.WriteRune('{')
+	if item.DocumentID != "" {
+		w.buf.WriteString(`"_id":`)
+		w.aux = strconv.AppendQuote(w.aux, item.DocumentID)
+		w.buf.Write(w.aux)
+		w.aux = w.aux[:0]
+	}
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('}')
+	w.buf.WriteRune('\n')
+	return nil
+}
+
+// writeBody writes the item body to the buffer; it must be called under a lock.
+//
+func (w *worker) writeBody(item BulkIndexerItem) error {
+	// TODO(karmi): Handle errors
+	if item.Body != nil {
+		if _, err := w.buf.ReadFrom(item.Body); err != nil {
+			return err
+		}
+		w.buf.WriteRune('\n')
+	}
+	return nil
+}
+
+// flush writes out the worker buffer; it must be called under a lock.
+//
+func (w *worker) flush() error {
+	if w.isFlushing {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: Already flushing\n", w.id)
+		}
+		return nil
+	}
+
+	if w.buf.Len() < 1 {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: Buffer empty\n", w.id)
+		}
+		return nil
+	}
+
+	var (
+		err error
+		blk BulkIndexerResponse
+		ctx = context.Background()
+	)
+
+	defer func() {
+		w.items = w.items[:0]
+		w.buf.Reset()
+		w.isFlushing = false
+	}()
+
+	w.isFlushing = true
+
+	if w.bi.config.DebugLogger != nil {
+		w.bi.config.DebugLogger.Printf("[worker-%03d] Flush: %s\n", w.id, w.buf.String())
+	}
+
+	atomic.AddUint64(&w.bi.stats.numRequests, 1)
+	req := esapi.BulkRequest{
+		Index: w.bi.config.Index,
+		Body:  w.buf,
+
+		Pipeline:            w.bi.config.Pipeline,
+		Refresh:             w.bi.config.Refresh,
+		Routing:             w.bi.config.Routing,
+		Source:              w.bi.config.Source,
+		SourceExcludes:      w.bi.config.SourceExcludes,
+		SourceIncludes:      w.bi.config.SourceIncludes,
+		Timeout:             w.bi.config.Timeout,
+		WaitForActiveShards: w.bi.config.WaitForActiveShards,
+
+		Pretty:     w.bi.config.Pretty,
+		Human:      w.bi.config.Human,
+		ErrorTrace: w.bi.config.ErrorTrace,
+		FilterPath: w.bi.config.FilterPath,
+		Header:     w.bi.config.Header,
+	}
+	res, err := req.Do(ctx, w.bi.config.Client)
+	if err != nil {
+		atomic.AddUint64(&w.bi.stats.numFailed, uint64(len(w.items)))
+		// TODO(karmi): Wrap error
+		return fmt.Errorf("flush: %s", err)
+	}
+	if res.Body != nil {
+		defer res.Body.Close()
+	}
+	if res.IsError() {
+		atomic.AddUint64(&w.bi.stats.numFailed, uint64(len(w.items)))
+		// TODO(karmi): Wrap error (include response struct)
+		return fmt.Errorf("flush: %s", res.String())
+	}
+
+	if err := w.bi.config.Decoder.UnmarshalFromReader(res.Body, &blk); err != nil {
+		// TODO(karmi): Wrap error (include response struct)
+		return fmt.Errorf("flush: error parsing response body: %s", err)
+	}
+
+	for i, blkItem := range blk.Items {
+		var (
+			item BulkIndexerItem
+			info BulkIndexerResponseItem
+			op   string
+		)
+
+		item = w.items[i]
+		for k, v := range blkItem {
+			op = k
+			info = v
+		}
+		if info.Error.Type != "" || info.Status > 201 {
```

Comment:
Why are you using `201` here? The `BulkeRequest.Do` returns `true` for `response.isError()` when `statusCode >= 299` which I think is more correct.

Reply:
Note that this is a status _for item_, part of the JSON payload in the response body, not the overall HTTP response — for a created document, the status code will be `201`.